File: chapter 3/3.1. Classification/3.1.2. Gaussian Naive Bayesian Classifier/gauss_naive_bayes_classifier.R
library(ggplot2)
library(Cairo)

# generate and plot training data

x_class_1 <- seq(2, 6, 0.2)
y_class_1 <- 1 + rnorm(length(x_class_1)) * 2

x_class_2 <- seq(14, 18, 0.2)
y_class_2 <- 1 + rnorm(length(x_class_2)) * 2

x_class_3 <- seq(7, 12, 0.2)
y_class_3 <- 14 + rnorm(length(x_class_3)) * 2

x <- c(x_class_1, x_class_2, x_class_3)
y <- c(y_class_1, y_class_2, y_class_3)

classes <- c(rep(1, length(x_class_1)), rep(2, length(x_class_2)), rep(3, length(x_class_3)))

train_data <- cbind(x, y, classes)
J <- 2

palette(rainbow(3))

CairoWin()

ggplot(data.frame(x, y), aes(x = x, y = y)) +
  geom_point(pch = c(14 + classes), size = c(rep(4, nrow(train_data))), col = c(classes)) +
  xlab('\nx1') +
  ylab('x2\n') +
  theme_bw() +
  theme(
    text = element_text(size = 25),
    axis.text = element_text(size = 25),
    plot.title = element_text(hjust = 0.5, size = 25)
  ) +
  ggtitle('Training data\n') +
  scale_x_continuous(breaks = seq(0, 20, 10), limits = c(0, 20)) +
  scale_y_continuous(breaks = seq(-10, 20, 10), limits = c(-10, 20))

ggsave(file = 'naive_bayes_train_data.png', type = 'cairo-png')

# learn a Gaussian Naive Bayesian Classifier

# prior P(Y)
prior <- function(data, K) {
  p <- rep(1, K)
  for(k in 1:K) {
    p[k] <- sum(data[, ncol(data)] == k) / nrow(data)
  }
  return(p)
}
prior <- prior(train_data, 3)

# learn J*K Gaussian components of the likelihood P(X|Y)
learn_gaussians <- function(train_data, J, K) {
  mean <- matrix(nrow = J, ncol = K)
  variance <- matrix(nrow = J, ncol = K)
  classes <- train_data[, ncol(train_data)]
  for(j in 1:J) {
    for(k in 1:K) {
      mean[j, k] <- sum(train_data[classes == k, j]) / sum(classes == k)
      variance[j, k] <- sum((train_data[classes == k, j] - mean[j, k])^2) / (sum(classes == k) - 1)
    }
  }
  return(list(mean, variance))
}
gaussians <- learn_gaussians(train_data, 2, 3)
gauss_mean <- gaussians[[1]]
gauss_variance <- gaussians[[2]]

# likelihood function for one data point xi, P(xi|Y=k)
likelihood <- function(xi, k, gauss_mean, gauss_variance) {
  product_of_gaussians <- 1
  for(j in 1:J) {
    gaussian_ <- dnorm(xi[j], gauss_mean[j, k], sqrt(gauss_variance[j, k]))
    product_of_gaussians <- product_of_gaussians * gaussian_
  }
  return(product_of_gaussians)
}

# compute posterior P(Y|xi) for one data point xi, for each class k
posterior <- function(xi, K, gauss_mean, gauss_variance) {
  post <- rep(0, K)
  for(k in 1:K) {
    post[k] <- prior[k] * likelihood(xi, k, gauss_mean, gauss_variance)
  }
  return(post)
}

# generate test data
x_test <- seq(0, 20, 0.2)
y_test <- 3 + rnorm(length(x_test)) * 12
data_test <- cbind(x_test, y_test)

# classify with Naive Bayes
classes_test <- rep(1, length(x_test))
for(i in 1:nrow(data_test)) {
  classes_test[i] <- which.max(posterior(data_test[i, ], 3, gauss_mean, gauss_variance))
}

CairoWin()

ggplot(data.frame(x_test, y_test), aes(x = x_test, y = y_test)) +
  geom_point(pch = c(14 + classes_test), size = c(rep(4, length(x_test))), col = c(classes_test)) +
  xlab('\nx1') +
  ylab('x2\n') +
  theme_bw() +
  theme(
    text = element_text(size = 25),
    axis.text = element_text(size = 25),
    plot.title = element_text(hjust = 0.5, size = 25)
  ) +
  ggtitle('Test data\n') +
  scale_x_continuous(breaks = seq(0, 20, 10), limits = c(0, 20)) +
  scale_y_continuous(breaks = seq(-10, 20, 10), limits = c(-10, 20))

ggsave(file = 'naive_bayes_test_data.png', type = 'cairo-png')

File: chapter 3/Python code/gauss_naive_bayes_classifier.py
import numpy as np
import matplotlib.pyplot as plt
import scipy.stats

# generate training data

x_class_0 = np.arange(2, 6, 0.2)
y_class_0 = 1 + (np.random.normal(0, 1, len(x_class_0))) * 2

x_class_1 = np.arange(14, 18, 0.2)
y_class_1 = 1 + (np.random.normal(0, 1, len(x_class_1))) * 2

x_class_2 = np.arange(7, 12, 0.2)
y_class_2 = 14 + (np.random.normal(0, 1, len(x_class_2))) * 2

x = np.concatenate([x_class_0, x_class_1, x_class_2])
y = np.concatenate([y_class_0, y_class_1, y_class_2])

classes_0 = np.zeros((len(x_class_0)))
classes_1 = np.zeros((len(x_class_1))) + 1
classes_2 = np.zeros((len(x_class_2))) + 2

classes = np.concatenate([classes_0, classes_1, classes_2])

train_data = np.column_stack((x,y,classes))

colors = ['r', 'g', 'b']
f = lambda x: colors[int(x)]
colors_train = list(map(f, classes))

# plot training data

my_dpi = 96
plt.figure(figsize=(800/my_dpi, 800/my_dpi), dpi=my_dpi)

plt.scatter(x,y, color=colors_train, s=20)

plt.xlabel('X')
plt.ylabel('Y')
plt.title('Training Data\n')

plt.savefig("gauss_naive_bayes_classifier_train.png")



# learn a Gaussian Naive Bayesian Classifier

# prior P(Y)
def prior(data, K):
    p = np.ones((K))
    nrow = data.shape[0]
    for k in np.arange(K):
        p[k] = np.sum(data[:, -1] == k) / nrow
    return p
prior_ = prior(train_data, 3)

# learn J*K Gaussian components of the likelihood P(X|Y)
J = 2
def learn_gaussians(train_data, J, K):
    mean = np.zeros(J*K).reshape((J,K))
    variance = np.zeros(J*K).reshape((J,K))
    classes = train_data[:, -1]    
    for j in np.arange(J):
        for k in np.arange(K):
            mean[j,k] = np.sum(train_data[classes==k,j]) / np.sum(classes == k)
            variance[j,k] = np.sum((train_data[classes==k,j] - mean[j,k])**2) / (np.sum(classes == k) - 1)  
    return mean, variance

gaussians = learn_gaussians(train_data, 2, 3)
gauss_mean = gaussians[0]
gauss_variance = gaussians[1]

# likelihood function for one data point xi, P(xi|Y=k)
def likelihood(xi, k, gauss_mean, gauss_variance):
    product_of_gaussians = 1
    for j in np.arange(J):
        gaussian_ = scipy.stats.norm.pdf(xi[j], gauss_mean[j,k], np.sqrt(gauss_variance[j,k]))
        product_of_gaussians = product_of_gaussians * gaussian_    
    return product_of_gaussians

# compute posterior P(Y|xi) for one data point xi, for each class k
def posterior(xi, K, gauss_mean, gauss_variance):
    post = np.zeros((K))
    for k in np.arange(K):
        post[k] = prior_[k] * likelihood(xi, k, gauss_mean, gauss_variance)     
    return post

# generate test data
x_test = np.arange(0, 20, 0.2)
y_test = 3 + (np.random.normal(0, 1, len(x_test))) * 12
data_test = np.column_stack((x_test,y_test))

# classify with Naive Bayes
classes_test = np.ones((len(x_test)))
for i in np.arange(data_test.shape[0]):
    post = posterior(data_test[i,:], 3, gauss_mean, gauss_variance)
    classes_test[i] = (np.where(post == np.max(post)))[0][0]

colors_test = list(map(f, classes_test))

# plot test data

plt.figure(figsize=(800/my_dpi, 800/my_dpi), dpi=my_dpi)

plt.scatter(x_test,y_test, color=colors_test, s=20)

plt.xlabel('X')
plt.ylabel('Y')
plt.title('Test Data\n')

plt.savefig("gauss_naive_bayes_classifier_test.png")
